images/keycloak-operator/tests/keycloak-test.sh

```diff
@@ -2,13 +2,276 @@
+
 TMPDIR="$(mktemp -d)"
 NAMESPACE="keycloak-test"
-apk add openssl
+REQUEST_RETRIES=5
+RETRY_DELAY=10
+ 
+apk add jq curl openssl openssl-config yq
 
-# Create Namespace and CRDs
+# Function to test Keycloak API
+TEST_keycloak_api() {
+  local attempt=0
+  local success=false
+
+  # Retrieve the initial admin credentials
+  local admin_username=$(kubectl get secret example-kc-initial-admin -n ${NAMESPACE} -o jsonpath='{.data.username}' | base64 -d)
+  local admin_password=$(kubectl get secret example-kc-initial-admin -n ${NAMESPACE} -o jsonpath='{.data.password}' | base64 -d)
+
+  while (( attempt < ${REQUEST_RETRIES} )); do
+    # Get an API token for Keycloak.
+    local response=$(
+      curl --http1.1 -k -w "\nHTTP_STATUS_CODE:%{http_code}\n" \
+        -X POST \
+        "${KEYCLOAK_URL}/realms/master/protocol/openid-connect/token" \
+        --user "admin-cli:Psip5UvTO1EXUEwzb15nxLWnwdU1Nlcg" \
+        -H "content-type: application/x-www-form-urlencoded" \
+        -d "username=${admin_username}" \
+        -d "password=${admin_password}" \
+        -d "grant_type=password"
+    )
+
+    local response_code=$(echo "${response}" | grep "HTTP_STATUS_CODE:" | cut -d':' -f2)
+    local content=$(echo "${response}" | sed '/HTTP_STATUS_CODE/d')
+
+    if [[ "${response_code}" == "200" ]]; then
+      local access_token=$(echo "${content}" | jq --raw-output '.access_token')
+
+      # Invoke Keycloak's API to retrieve a list of users.
+      local users_output=$(curl --http1.1 -kv -X GET "${KEYCLOAK_URL}/admin/realms/master/users" \
+        -H "Authorization: Bearer ${access_token}")
+      success=true
+      break
+    else
+      sleep ${RETRY_DELAY}
+    fi
+    attempt=$((attempt+1))
+  done
+
+  if ! $success; then
+    echo "ERROR: Failed to get user data from the API after ${REQUEST_RETRIES} attempts." >&2
+    return 1
+  fi
+
+  # Ensure that an 'admin' user was returned in the API response.
+  extracted_username=$(echo "${users_output}" | jq -r '.[] | select(.username=="admin") | .username')
+  if [[ "${extracted_username}" == "admin" ]]; then
+    echo "Keycloak API correctly returned 'admin' user details."
+  else
+    echo "FAILED: No entry with username 'admin' found in the response: ${users_output}"
+    exit 1
+  fi
+}
+
+# Function to create a new realm if it doesn't exist
+create_realm_if_not_exists() {
+  local realm_name="test-realm"
+  local admin_username=$(kubectl get secret example-kc-initial-admin -n ${NAMESPACE} -o jsonpath='{.data.username}' | base64 -d)
+  local admin_password=$(kubectl get secret example-kc-initial-admin -n ${NAMESPACE} -o jsonpath='{.data.password}' | base64 -d)
+
+  # Get an API token for Keycloak.
+  local token_response
+  token_response=$(curl --http1.1 -k -sS -w "\nHTTP_STATUS_CODE:%{http_code}\n" \
+    -X POST "${KEYCLOAK_URL}/realms/master/protocol/openid-connect/token" \
+    --user "admin-cli:Psip5UvTO1EXUEwzb15nxLWnwdU1Nlcg" \
```

Comment:
These "Psip5UvTO1EXUEwzb15nxLWnwdU1Nlcg" strings look like passwords, but aren't. They also aren't required. There are a couple of instances.

I'd recommend removing them to avoid confusion:

```
--user "admin-cli:"
```
(You need the final : or curl prompts you for the password)